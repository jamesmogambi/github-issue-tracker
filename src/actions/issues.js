require("isomorphic-fetch");

// SET_REPOSITORY_ISSUES
export const setRepositoryIssues = (payload) => ({
  type: "SET_REPOSITORY_ISSUES",
  payload,
});
// CLEAR_ISSUES
export const clearIssues = () => ({
  type: "CLEAR_ISSUES",
});
// ERROR
export const error = (payload) => ({
  type: "ERROR",
  payload,
});

// Get Github Repo Issues
export const getRepositoryIssues = (params) => {
  return (dispatch) => {
    dispatch(clearIssues());
    fetch("https://api.github.com/graphql", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        authorization: `bearer ${process.env.GITHUB_ACCESS_TOKEN}`,
      },

      body: JSON.stringify({
        query: `
                  query($repoOwner: String!, $repoName: String!) {
                  repository(name: $repoName, owner: $repoOwner) {
                    issues(first: 100) {
                      edges {
                        node {
                           id
                           number
                           state
                           createdAt
                           title
                           url
                           author{
                            avatarUrl
                            login
                          }
                           bodyHTML
                        }
                      }
                    }
                  }
                }
                 `,
        variables: {
          repoOwner: params.repoOwner,
          repoName: params.repoName,
        },
      }),
    })
      .then((res) => res.json())
      .then((res) => {
        if (res.data.repository) {
          dispatch(
            setRepositoryIssues({
              repositoryissues: res.data.repository.issues.edges,
              username: params.repoOwner,
              repositoryname: params.repoName,
            })
          );
        }
      })
      .catch((err) => {
        dispatch(error());
        console.error(err);
      });
  };
};
