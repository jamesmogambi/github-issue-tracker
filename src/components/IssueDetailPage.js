import React from "react";
import { connect } from "react-redux";

export const IssueDetailsPage = (props) => (
  <div>
    <div className="content-container">
      {props.issue ? (
        <div>
          <h1 className="page-header__title">
            Viewing details of Issue <span>{props.issue.node.title}</span>{" "}
          </h1>
          <div
            dangerouslySetInnerHTML={{ __html: props.issue.node.bodyHTML }}
          ></div>
        </div>
      ) : (
        <div className="list-item list-item--message">
          <span>No Details </span>
        </div>
      )}
    </div>
  </div>
);

const mapStateToProps = (state, props) => {
  return {
    issue: state.issues.repositoryissues.find(
      (issue) => issue.node.id === props.match.params.id
    ),
  };
};

export default connect(mapStateToProps)(IssueDetailsPage);
