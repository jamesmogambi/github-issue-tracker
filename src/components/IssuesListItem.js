import React from "react";
import { Link } from "react-router-dom";

const IssuesListItem = ({ id, title, author, createdAt }) => (
  <Link className="list-item" to={`/issueDetails/${id}`}>
    <div>
      <img src={author.avatarUrl} alt="Avatar" className="avatar" />
      <div className="list-item__sub-title">{title}</div>
    </div>
    <span className="list-item__sub-title ">opened at {createdAt}</span>
  </Link>
);

export default IssuesListItem;
