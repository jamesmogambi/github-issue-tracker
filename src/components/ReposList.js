import React from "react";
import { connect } from "react-redux";
import ReposListItem from "./ReposListItem";

export const ReposList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Repositories</div>
      <div className="show-for-desktop">Repository</div>
      <div className="show-for-desktop">Action</div>
    </div>
    <div className="list-body">
      {props.repos.repositories.length === 0 ? (
        <div className="list-item list-item--message">
          <span>No Repos</span>
        </div>
      ) : (
        props.repos.repositories.map((repo) => {
          return (
            <ReposListItem
              key={repo.id}
              repositoryOwner={props.repos.repositoryOwner}
              {...repo}
            />
          );
        })
      )}
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    repos: state.repos,
  };
};

export default connect(mapStateToProps)(ReposList);
