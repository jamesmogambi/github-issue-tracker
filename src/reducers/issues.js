// Issues Reducer

const initialState = {
  username: null,
  repositoryname: null,
  repositoryissues: [],
  error: null,
};

export default (state = initialState, action) => {
  switch (action.type) {
    case "SET_REPOSITORY_ISSUES":
      const { username, repositoryname, repositoryissues } = action.payload;
      return {
        ...state,
        username,
        repositoryname,
        repositoryissues,
      };

    case "CLEAR_ISSUES":
      return {
        ...state,
        username: null,
        repositoryname: null,
        repositoryissues: [],
      };
    default:
      return state;
  }
};
