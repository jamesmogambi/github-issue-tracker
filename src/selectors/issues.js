// Get visible issues
export default (issues, { text }) => {
  return issues.filter(({ node }) => {
    const textMatch = node.title.toLowerCase().includes(text.toLowerCase());
    return textMatch;
  });
};
