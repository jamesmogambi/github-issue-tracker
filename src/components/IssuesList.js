import React from "react";
import { connect } from "react-redux";
import { getRepositoryIssues } from "../actions/issues";
import IssuesListItem from "./IssuesListItem";
import selectIssues from "../selectors/issues";

// export class IssuesList extends React.Component {
//   constructor(props) {
//     super(props);
//   }
//   componentWillMount() {
//     this.props.getRepositoryIssues(this.props.params);
//   }
//   render() {
//     return (
//       <div className="content-container">
//         <h1 className="page-header__title">
//           Viewing Issues of Repository <span>{this.props.params.repoName}</span>{" "}
//         </h1>

//         <div className="list-header">
//           <div className="show-for-mobile">Issues</div>
//           <div className="show-for-desktop">Issue</div>
//           <div className="show-for-desktop">Created</div>
//         </div>
//         <div className="list-body">
//           {this.props.issues.length === 0 ? (
//             <div className="list-item list-item--message">
//               <span>No Issues for this Repo</span>
//             </div>
//           ) : (
//             this.props.issues.map(({ node }) => {
//               return <IssuesListItem key={node.id} {...node} />;
//             })
//           )}
//         </div>
//       </div>
//     );
//   }
// }

// const mapDispatchToProps = (dispatch) => ({
//   getRepositoryIssues: (params) => dispatch(getRepositoryIssues(params)),
// });

// const mapStateToProps = (state) => {
//   return {
//     issues: selectIssues(state.issues.repositoryissues, state.filters),
//   };
// };

// export default connect(mapStateToProps, mapDispatchToProps)(IssuesList);

export const IssuesList = (props) => (
  <div className="content-container">
    <h1 className="page-header__title">
      Viewing Issues of Repository <span>{props.repoName}</span>{" "}
    </h1>

    <div className="list-header">
      <div className="show-for-mobile">Issues</div>
      <div className="show-for-desktop">Issue</div>
      <div className="show-for-desktop">Created</div>
    </div>
    <div className="list-body">
      {props.issues.length === 0 ? (
        <div className="list-item list-item--message">
          <span>No Issues for this Repo</span>
        </div>
      ) : (
        props.issues.map(({ node }) => {
          return <IssuesListItem key={node.id} {...node} />;
        })
      )}
    </div>
  </div>
);

export default IssuesList;
