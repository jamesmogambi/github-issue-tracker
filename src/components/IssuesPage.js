import React from "react";
import { connect } from "react-redux";
import { getRepositoryIssues } from "../actions/issues";
import IssuesList from "./IssuesList";
import IssuesListFilters from "./IssuesListFilters";
import selectIssues from "../selectors/issues";

export class IssuesPage extends React.Component {
  constructor(props) {
    super(props);
  }
  componentWillMount() {
    this.props.getRepositoryIssues(this.props.match.params);
  }
  render() {
    return (
      <div>
        <IssuesListFilters />
        <IssuesList issues={this.props.issues} repoName={this.props.repoName} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  getRepositoryIssues: (params) => dispatch(getRepositoryIssues(params)),
});

const mapStateToProps = (state) => {
  return {
    issues: selectIssues(state.issues.repositoryissues, state.filters),
    repoName: state.issues.repositoryname,
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(IssuesPage);
