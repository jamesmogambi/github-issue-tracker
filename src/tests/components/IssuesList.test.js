import React from "react";
import { shallow } from "enzyme";
import { IssuesList } from "../../components/IssuesList";
import issues from "../fixtures/issues";

test("should render IssuesList with issues", () => {
  const data = {
    issues: issues,
    repoName: "James Edwards",
  };
  const wrapper = shallow(
    <IssuesList issues={data.issues} repoName={data.repoName} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("should render IssuesList with empty message", () => {
  const data = {
    issues: [],
    repoName: null,
  };
  const wrapper = shallow(
    <IssuesList issues={data.issues} repoName={data.repoName} />
  );
  expect(wrapper).toMatchSnapshot();
});
