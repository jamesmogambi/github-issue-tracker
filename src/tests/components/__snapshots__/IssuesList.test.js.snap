// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render IssuesList with empty message 1`] = `
<div
  className="content-container"
>
  <h1
    className="page-header__title"
  >
    Viewing Issues of Repository 
    <span />
     
  </h1>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Issues
    </div>
    <div
      className="show-for-desktop"
    >
      Issue
    </div>
    <div
      className="show-for-desktop"
    >
      Created
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No Issues for this Repo
      </span>
    </div>
  </div>
</div>
`;

exports[`should render IssuesList with issues 1`] = `
<div
  className="content-container"
>
  <h1
    className="page-header__title"
  >
    Viewing Issues of Repository 
    <span>
      James Edwards
    </span>
     
  </h1>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Issues
    </div>
    <div
      className="show-for-desktop"
    >
      Issue
    </div>
    <div
      className="show-for-desktop"
    >
      Created
    </div>
  </div>
  <div
    className="list-body"
  >
    <IssuesListItem
      author={
        Object {
          "avatarUrl": "https://avatars.githubusercontent.com/u/31744209?u=a6d48cf8f85179ba4db1b3268e8624fd21ddc2e5&v=4",
          "login": "jamesmogambi",
        }
      }
      bodyHTML="<p><a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://user-images.githubusercontent.com/31744209/102247600-34eeef00-3f11-11eb-8c83-9b99b0ace69b.png\\"><img src=\\"https://user-images.githubusercontent.com/31744209/102247600-34eeef00-3f11-11eb-8c83-9b99b0ace69b.png\\" alt=\\"profiles\\" style=\\"max-width:100%;\\"></a><br>
    <a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://user-images.githubusercontent.com/31744209/102247609-37514900-3f11-11eb-8366-1af9b2f0a6cb.png\\"><img src=\\"https://user-images.githubusercontent.com/31744209/102247609-37514900-3f11-11eb-8366-1af9b2f0a6cb.png\\" alt=\\"sign-in\\" style=\\"max-width:100%;\\"></a><br>
    <a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://user-images.githubusercontent.com/31744209/102247613-37514900-3f11-11eb-9a0f-8e1a1b8d3d22.png\\"><img src=\\"https://user-images.githubusercontent.com/31744209/102247613-37514900-3f11-11eb-9a0f-8e1a1b8d3d22.png\\" alt=\\"user-dashboard\\" style=\\"max-width:100%;\\"></a><br>
    <a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://user-images.githubusercontent.com/31744209/102247619-38827600-3f11-11eb-8404-a310eb6663d1.png\\"><img src=\\"https://user-images.githubusercontent.com/31744209/102247619-38827600-3f11-11eb-8404-a310eb6663d1.png\\" alt=\\"comment\\" style=\\"max-width:100%;\\"></a><br>
    <a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://user-images.githubusercontent.com/31744209/102248587-5e5c4a80-3f12-11eb-845a-b99de1197453.png\\"><img src=\\"https://user-images.githubusercontent.com/31744209/102248587-5e5c4a80-3f12-11eb-845a-b99de1197453.png\\" alt=\\"edit-profile\\" style=\\"max-width:100%;\\"></a><br>
    <a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://user-images.githubusercontent.com/31744209/102247624-391b0c80-3f11-11eb-8863-60c4a5edcff6.png\\"><img src=\\"https://user-images.githubusercontent.com/31744209/102247624-391b0c80-3f11-11eb-8863-60c4a5edcff6.png\\" alt=\\"index\\" style=\\"max-width:100%;\\"></a><br>
    <a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://user-images.githubusercontent.com/31744209/102247631-3a4c3980-3f11-11eb-8feb-6e50c76e65a6.png\\"><img src=\\"https://user-images.githubusercontent.com/31744209/102247631-3a4c3980-3f11-11eb-8feb-6e50c76e65a6.png\\" alt=\\"posts\\" style=\\"max-width:100%;\\"></a></p>"
      createdAt="2020-12-15T17:20:41Z"
      id="MDU6SXNzdWU3Njc4NjI5Nzk="
      number={1}
      state="OPEN"
      title="screenshots"
      url="https://github.com/jamesmogambi/dev-_MERN-Stack/issues/1"
    />
    <IssuesListItem
      author={
        Object {
          "avatarUrl": "https://avatars.githubusercontent.com/u/31744209?u=a6d48cf8f85179ba4db1b3268e8624fd21ddc2e5&v=4",
          "login": "jamesmogambi",
        }
      }
      bodyHTML="<p><a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://user-images.githubusercontent.com/31744209/107021668-800cdc00-67b5-11eb-931e-89740b2a4cae.jpg\\"><img src=\\"https://user-images.githubusercontent.com/31744209/107021668-800cdc00-67b5-11eb-931e-89740b2a4cae.jpg\\" alt=\\"IMG_20210119_084748\\" style=\\"max-width:100%;\\"></a></p>"
      createdAt="2021-02-05T10:43:31Z"
      id="MDU6SXNzdWU4MDIwNTc4MDI="
      number={2}
      state="CLOSED"
      title="author"
      url="https://github.com/jamesmogambi/dev-_MERN-Stack/issues/2"
    />
  </div>
</div>
`;
